File: objc/objcsearch/Tests/objcsearchTests/SearcherTests.m
//
//  SearcherTest.m
//  objcsearch_tests
//
//  Created by Cary Clark on 11/10/18.
//  Copyright © 2018 Cary Clark. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "FileUtil.h"
#import "SearchConfig.h"
#import "Searcher.h"
#import "SearchSettings.h"

@interface SearcherTests : XCTestCase
@property SearchConfig *config;
@property NSString *testFilePath;
@end

@implementation SearcherTests

- (void)setUp {
    [super setUp];
    self.config = [[SearchConfig alloc] init];
    self.testFilePath = [NSString stringWithFormat:@"%@/shared/testFiles/testFile2.txt", self.config.xsearchPath];
}

- (void)tearDown {
    [super tearDown];
}

/*************************************************************
 * searchMultiLineString tests
 *************************************************************/
- (void)testSearchMultiLineString {
    SearchSettings *settings = [[SearchSettings alloc] init];
    [settings addPath:self.testFilePath];
    [settings addSearchPattern:@"Searcher"];
    NSError *error = nil;
    Searcher *searcher = [[Searcher alloc] initWithConfig:self.config settings:settings error:&error];
    
    NSArray<SearchResult*> *results = [searcher search:&error];

    XCTAssert([results count] == 2);

    SearchResult *firstResult = [results objectAtIndex:0];
    int expectedFirstLineNum = 30;
    XCTAssert([firstResult lineNum] == expectedFirstLineNum);
    int expectedFirstMatchStartIndex = 3;
    XCTAssert([firstResult matchStartIndex] == expectedFirstMatchStartIndex);
    int expectedFirstMatchEndIndex = 11;
    XCTAssert([firstResult matchEndIndex] == expectedFirstMatchEndIndex);

    SearchResult *secondResult = [results objectAtIndex:1];
    int expectedSecondLineNum = 36;
    XCTAssert([secondResult lineNum] == expectedSecondLineNum);
    int expectedSecondMatchStartIndex = 24;
    XCTAssert([secondResult matchStartIndex] == expectedSecondMatchStartIndex);
    int expectedSecondMatchEndIndex = 32;
    XCTAssert([secondResult matchEndIndex] == expectedSecondMatchEndIndex);
}

@end

File: swift/swiftsearch/Tests/swiftsearchTests/SearcherTests.swift
import Cocoa
import XCTest

import swiftsearch
import swiftfind

class SearcherTests: XCTestCase {
    let config = SearchConfig()
    let fileTypes = FileTypes()

    func getSettings() -> SearchSettings {
        let settings = SearchSettings()
        settings.addPath(".")
        settings.addSearchPattern("Searcher")
        return settings
    }

    override func setUp() {
        super.setUp()
    }

    override func tearDown() {
        super.tearDown()
    }

    /* ==========================================================================
     * searchLineReader tests
     ========================================================================= */
    func testSearchLineReader() {
        let settings = getSettings()
        let searcher = try! Searcher(settings: settings)
        let testFilePath = FileUtil.joinPath(config.sharedPath, childPath: "testFiles/testFile2.txt")

        if let reader = StreamReader(path: testFilePath, encoding: .utf8) {
            let results = searcher.searchLineReader(reader)

            XCTAssert(results.count == 2)

            XCTAssertEqual(29, results[0].lineNum)
            XCTAssertEqual(3, results[0].matchStartIndex)
            XCTAssertEqual(11, results[0].matchEndIndex)

            XCTAssertEqual(35, results[1].lineNum)
            XCTAssertEqual(24, results[1].matchStartIndex)
            XCTAssertEqual(32, results[1].matchEndIndex)

            reader.close()

        } else {
            XCTAssertTrue(false)
        }
    }

    /* ==========================================================================
     * searchMultiLineString tests
     ========================================================================= */
    func testSearchMultiLineString() {
        let settings = getSettings()
        settings.multiLineSearch = true
        let searcher = try! Searcher(settings: settings)
        let testFilePath = FileUtil.joinPath(config.sharedPath, childPath: "testFiles/testFile2.txt")
        let testFileContents = try? String(contentsOfFile: testFilePath, encoding: .utf8)
        if testFileContents != nil {
            let results = searcher.searchMultiLineString(testFileContents!)

            XCTAssert(results.count == 2)

            XCTAssertEqual(29, results[0].lineNum)
            XCTAssertEqual(3, results[0].matchStartIndex)
            XCTAssertEqual(11, results[0].matchEndIndex)

            XCTAssertEqual(35, results[1].lineNum)
            XCTAssertEqual(24, results[1].matchStartIndex)
            XCTAssertEqual(32, results[1].matchEndIndex)

        } else {
            XCTAssertTrue(false)
        }
    }
}
